community/kernel/src/test/java/org/neo4j/graphdb/GraphDatabaseServiceTest.java

```diff
@@ -172,81 +173,44 @@ public void terminateNestedTransactionThrowsExceptionOnNextNestedOperation() thr
+        GraphDatabaseService db = new TestGraphDatabaseFactory().newImpermanentDatabase();
 
         // When
-        final CountDownLatch shutdown = new CountDownLatch( 1 );
-        final AtomicReference<Object> result = new AtomicReference<>();
-        Executors.newSingleThreadExecutor().submit( new Runnable()
+        Barrier.Control barrier = new Barrier.Control();
+        t2.execute( state ->
         {
-            @Override
-            public void run()
+            try ( Transaction tx = db.beginTx() )
             {
-                try ( Transaction tx = db.beginTx() )
-                {
-                    shutdown.countDown();
-
-                    try
-                    {
-                        Thread.sleep( 2000 );
-                    }
-                    catch ( InterruptedException e )
-                    {
-                        throw new RuntimeException( e );
-                    }
-
-                    db.createNode();
-                    tx.success();
-
-                    Executors.newSingleThreadExecutor().submit( new Runnable()
-                    {
-                        @Override
-                        public void run()
-                        {
-                            try
-                            {
-                                db.beginTx();
-                                result.set( Boolean.TRUE );
-                            }
-                            catch ( Exception e )
-                            {
-                                result.set( e );
-                            }
-
-                            synchronized ( result )
-                            {
-                                result.notifyAll();
-                            }
-                        }
-                    } );
-                }
-                catch ( Throwable e )
-                {
-                    throw new RuntimeException( e );
-                }
+                barrier.reached(); // <-- this triggers t3 to start a db.shutdown()
             }
+            return null;
         } );
 
-        shutdown.await();
-        db.shutdown();
+        barrier.await();
+        Future<Object> shutdownFuture = t3.execute( state ->
+        {
+            db.shutdown();
+            return null;
+        } );
+        t3.get().waitUntilWaiting( location -> location.isAt( DatabaseAvailability.class, "shutdown" ) );
+        barrier.release(); // <-- this triggers t2 to continue its transaction
+        shutdownFuture.get();
 
-        while ( result.get() == null )
+        try
         {
-            synchronized ( result )
-            {
-                result.wait( 100 );
-            }
+            db.beginTx();
+            fail( "Should fail" );
+        }
+        catch ( DatabaseShutdownException e )
+        {
+            //THEN good
         }
-
-        assertThat( result.get(), instanceOf( DatabaseShutdownException.class ) );
     }
```

Comment:
My understanding from the test name is that this is intended to test the case where a transaction starts while the shutdown is in progress. So I would expect the `beginTx()` to come before the `barrier.release()`. Otherwise I don't see any reason to have the blocking tx in this test -- you  might as well just do a simple `db.shutdown()` followed by `db.beginTx()`.